packages/commonwealth/server/routes/bulkBalances.ts

```diff
@@ -0,0 +1,172 @@
+import TokenBalanceCache from 'token-balance-cache/src/index';
+import { AddressInstance } from 'server/models/address';
+import { QueryTypes } from 'sequelize';
+import { DB, sequelize } from '../database';
+import { AppError, ServerError } from '../util/errors';
+import { success, TypedRequestBody, TypedResponse } from '../types';
+
+enum BulkBalancesErrors {
+  InvalidToken = 'Invalid token',
+  NoProfileId = 'No profileId provided',
+  NoProfile = "Profile doesn't exist",
+  NoRegisteredAddresses = 'No wallet addresses registered to this profile',
+  Failed = 'Request Failed',
+}
+
+// ----------- OUTSTANDING QUESTIONS -----------
+// 1. Max across addresses or sum across addresses?
+// 2. How is this route being authenticated, if at all?
+
+type bulkBalanceReq = {
+  profileId: number;
+  token: string;
+  chainNodes: {
+    [nodeId: number]: string | string[];
+  };
+};
+
+type bulkBalanceResp = {
+  balances: {
+    [nodeId: string]:
+      | {
+          [tokenAddress: string]: number;
+        }
+      | number;
+  };
+  bases: string[];
+};
+
+const bulkBalances = async (
+  models: DB,
+  tokenBalanceCache: TokenBalanceCache,
+  req: TypedRequestBody<bulkBalanceReq>,
+  res: TypedResponse<bulkBalanceResp>
+) => {
+  const { profileId, chainNodes, token } = req.body;
+
+  if (!process.env.DISCORD_BOT_TOKEN)
+    throw new AppError(BulkBalancesErrors.Failed);
+
+  if (!token || token !== process.env.DISCORD_BOT_TOKEN)
+    throw new AppError(BulkBalancesErrors.InvalidToken);
+
+  if (!profileId) throw new AppError(BulkBalancesErrors.NoProfileId);
+
+  const profile = await models.Profile.findOne({
+    where: { id: profileId },
+  });
+  if (!profile) throw new ServerError(BulkBalancesErrors.NoProfile);
```

Comment:
This should be an AppError as it's related to validating user input.

Reply:
resolved